File: sample-kotlin-compiler/src/main/java/com/redmadrobot/samplekotlincompiler/SampleCompiler.java
package com.redmadrobot.samplekotlincompiler;

import com.google.auto.service.AutoService;

import com.redmadrobot.samplelib.SampleAnnotation;

import java.io.File;
import java.io.IOException;
import java.io.Writer;
import java.lang.annotation.Annotation;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import javax.annotation.processing.AbstractProcessor;
import javax.annotation.processing.Messager;
import javax.annotation.processing.ProcessingEnvironment;
import javax.annotation.processing.Processor;
import javax.annotation.processing.RoundEnvironment;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.Element;
import javax.lang.model.element.ElementKind;
import javax.lang.model.element.TypeElement;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;

import static javax.lang.model.SourceVersion.latestSupported;
import static javax.tools.StandardLocation.SOURCE_OUTPUT;

@SuppressWarnings("unused")
@AutoService(Processor.class)
public class SampleCompiler extends AbstractProcessor {

    private Messager messager;

    private Types typeUtils;

    private Elements elementUtils;

    @Override
    public synchronized void init(ProcessingEnvironment processingEnv) {
        super.init(processingEnv);

        messager = processingEnv.getMessager();
        typeUtils = processingEnv.getTypeUtils();
        elementUtils = processingEnv.getElementUtils();
    }

    @Override
    public Set<String> getSupportedAnnotationTypes() {
        Set<String> supportedAnnotationTypes = new HashSet<>();
        Collections.addAll(supportedAnnotationTypes, SampleAnnotation.class.getCanonicalName());
        return supportedAnnotationTypes;
    }

    @Override
    public SourceVersion getSupportedSourceVersion() {
        return latestSupported();
    }

    @Override
    public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {

        processInjectors(roundEnv, SampleAnnotation.class, ElementKind.CLASS,
                new SampleClassGenerator());

        return true;
    }

    private void processInjectors(final RoundEnvironment roundEnv,
            Class<? extends Annotation> clazz, ElementKind kind, ClassGenerator classGenerator) {
        for (Element annotatedElements : roundEnv.getElementsAnnotatedWith(clazz)) {
            if (annotatedElements.getKind() != kind) {
                throw new RuntimeException(
                        annotatedElements + " must be " + kind.name() + ", or not mark it as @"
                                + clazz.getSimpleName());
            }

            generateCode(kind, classGenerator, annotatedElements);
        }
    }

    private void generateCode(ElementKind kind, ClassGenerator classGenerator, Element element) {
        if (element.getKind() != kind) {
            throw new RuntimeException(element + " must be " + kind.name());
        }

        List<ClassGeneratingParams> classGeneratingParamsList = new ArrayList<>();

        //noinspection unchecked
        final boolean generated = classGenerator.generate(element, classGeneratingParamsList);

        if (!generated) {
            return;
        }

        for (ClassGeneratingParams classGeneratingParams : classGeneratingParamsList) {
            createSourceFile(classGeneratingParams);
        }
    }

    private void createSourceFile(ClassGeneratingParams classGeneratingParams) {

        Map<String, String> options = processingEnv.getOptions();
        String generatedPath = options.get("kapt.kotlin.generated");
        String path = generatedPath
                .replaceAll("(.*)tmp(/kapt/debug/)kotlinGenerated",
                        "$1generated/source$2");

        String packageName = generatedPath
                .replaceAll("(.*)tmp(/kapt/debug/)kotlinGenerated",
                        "$1generated/source$2");


				/*File srcFile = new File(path + classGeneratingParams.getName()
                                                .replace(".", "/") + ".java");
				*/

        String destFileFullClass = path + classGeneratingParams.getName()
                .replace(".", "/") + ".kt";
        File destFile = new File(destFileFullClass);

        try {
            Writer w = processingEnv.getFiler()
                    .createResource(SOURCE_OUTPUT, classGeneratingParams.getPackageNme(),
                            classGeneratingParams.getSimpleFileName())
                    .openWriter();

            w.write(classGeneratingParams.getBody());
            w.flush();
            w.close();


        } catch (IOException e) {
            e.printStackTrace();
        }


    }
}

File: sample-kotlin-compiler/src/main/java/com/redmadrobot/samplekotlincompiler/SampleCompiler.kt
package com.redmadrobot.samplekotlincompiler

import com.redmadrobot.samplelib.SampleAnnotation
import java.io.File
import java.io.IOException
import java.util.*
import javax.annotation.processing.*
import javax.lang.model.SourceVersion
import javax.lang.model.SourceVersion.latestSupported
import javax.lang.model.element.*
import javax.lang.model.util.Elements
import javax.lang.model.util.Types
import javax.tools.StandardLocation.SOURCE_OUTPUT


class SampleCompiler : Processor {
    private lateinit var processingEnvironment: ProcessingEnvironment

    override fun getSupportedOptions(): MutableSet<String> {
        return mutableSetOf()
    }

    override fun getCompletions(element: Element?, annotation: AnnotationMirror?, member: ExecutableElement?, userText: String?): MutableIterable<Completion> {
        return mutableListOf()
    }

    private var messager: Messager? = null

    private var typeUtils: Types? = null

    private var elementUtils: Elements? = null

    @Synchronized override fun init(processingEnv: ProcessingEnvironment) {
        processingEnvironment = processingEnv
        messager = processingEnvironment.messager
        typeUtils = processingEnvironment.typeUtils
        elementUtils = processingEnvironment.elementUtils
    }

    override fun getSupportedAnnotationTypes(): Set<String> {
        val supportedAnnotationTypes = HashSet<String>()
        Collections.addAll(supportedAnnotationTypes, SampleAnnotation::class.java.canonicalName)
        return supportedAnnotationTypes
    }

    override fun getSupportedSourceVersion(): SourceVersion {
        return latestSupported()
    }

    override fun process(annotations: Set<TypeElement>, roundEnv: RoundEnvironment): Boolean {
        processInjectors(roundEnv, SampleAnnotation::class.java, ElementKind.CLASS,
                SampleClassGenerator())

        return true
    }

    private fun processInjectors(roundEnv: RoundEnvironment,
                                 clazz: Class<out Annotation>, kind: ElementKind, classGenerator: ClassGenerator<TypeElement>) {
        for (annotatedElements in roundEnv.getElementsAnnotatedWith(clazz)) {
            if (annotatedElements.kind != kind) {
                throw RuntimeException(
                        "$annotatedElements  must be  ${kind.name}  or not mark it as @${clazz.simpleName}")
            }


            generateCode(kind, classGenerator, annotatedElements as TypeElement)
        }
    }

    private fun generateCode(kind: ElementKind, classGenerator: ClassGenerator<TypeElement>, element: TypeElement) {
        if (element.kind != kind) {
            throw RuntimeException("$element  must be ${kind.name}")
        }

        val classGeneratingParamsList = ArrayList<ClassGeneratingParams>()

        //noinspection unchecked
        val generated = classGenerator.generate(element, classGeneratingParamsList)

        if (!generated) {
            return
        }

        for (classGeneratingParams in classGeneratingParamsList) {
            createSourceFile(classGeneratingParams)
        }
    }

    private fun createSourceFile(classGeneratingParams: ClassGeneratingParams) {

        val options = processingEnvironment.options
        val GENERATED_CLASS_PATH_OPTION = "kapt.kotlin.generated"

        val generatedPath = options[GENERATED_CLASS_PATH_OPTION] ?: throw RuntimeException("compatible with kapt only. kapt2 is not supported now")

        val path = generatedPath.replace("(.*)tmp(/kapt/debug/)kotlinGenerated".toRegex(), "$1generated/source$2")

        val packageName = generatedPath.replace("(.*)tmp(/kapt/debug/)kotlinGenerated".toRegex(), "$1generated/source$2")


        /*File srcFile = new File(path + classGeneratingParams.getName()
                                                .replace(".", "/") + ".java");
				*/

        val destFileFullClass = path + classGeneratingParams.name.replace(".", "/") + ".kt"
        val destFile = File(destFileFullClass)

        try {
            val w = processingEnvironment.filer.createResource(SOURCE_OUTPUT, classGeneratingParams.packageNme,
                    classGeneratingParams.getSimpleFileName()).openWriter()

            w.write(classGeneratingParams.body)
            w.flush()
            w.close()


        } catch (e: IOException) {
            e.printStackTrace()
        }


    }
}
